lstchain/reco/reconstructor.py

```diff
@@ -0,0 +1,977 @@
+        self.focal_length = subarray.tel[self.telescope_id].optics.equivalent_focal_length
+        readout = self.subarray.tel[self.telescope_id].camera.readout
+        sampling_rate = readout.sampling_rate.to_value(u.GHz)
+        self.dt = (1.0 / sampling_rate)
+        self.labels = {'charge': 'Charge [p.e.]',
+                       't_cm': '$t_{CM}$ [ns]',
+                       'x_cm': '$x_{CM}$ [m]',
+                       'y_cm': '$y_{CM}$ [m]',
+                       'length': r'$\sigma_l$ [m]',
+                       'wl': r'$\sigma_w$ / $\sigma_l$',
+                       'psi': r'$\psi$ [rad]',
+                       'v': '$v$ [m/ns]',
+                       'rl': 'length asymmetry'
+                       }
+        self.names_parameters = list(inspect.signature(self.log_pdf).parameters)
+        self.template = NormalizedPulseTemplate.load_from_eventsource(subarray.tel[1].camera.readout)
+        self.template_time_of_max = self.template.compute_time_of_max()
+        if self.use_interleaved is None:  # test to include interleaved correction on pedestal, NOT FUNCTIONAL
+            self.pedestal_std = None
+        else:
+            _, self.pedestal_std = get_bias_and_std(config['lh_fit_config']['use_interleaved'])
+        self.end_parameters = None
+        self.error_parameters = None
+        self.fcn = None
+
+        poisson_peaks = np.arange(self.n_peaks, dtype=int)
+        poisson_peaks[0] = 1
+        log_factorial = np.log(poisson_peaks)
+        self.log_factorial = np.cumsum(log_factorial)
+
+    def __call__(self, event, dl1_container):
+        self.event_id = event.index.event_id
+        self.image = event.dl1.tel[self.telescope_id].image
+        hillas_signal_pixels = event.dl1.tel[self.telescope_id].image_mask
+        start_x_cm, start_y_cm = init_centroid(dl1_container,
+                                               self.geometry[hillas_signal_pixels],
+                                               self.image[hillas_signal_pixels],
+                                               self.no_asymmetry
+                                               )
+
+        waveform = event.r1.tel[self.telescope_id].waveform
+
+        time_shift = None
+        if event.simulation is None:  # Find time shift correction for data
+            dl1_calib = event.calibration.tel[self.telescope_id].dl1
+            time_shift = dl1_calib.time_shift
+            if dl1_calib.pedestal_offset is not None:
+                waveform = waveform - dl1_calib.pedestal_offset[:, np.newaxis]
+
+        self.n_pixels, self.n_samples = waveform.shape
+        self.times = np.arange(0, self.n_samples) * self.dt
+        selected_gains = event.r1.tel[self.telescope_id].selected_gain_channel
+        self.is_high_gain = (selected_gains == 0)
+
+        v = dl1_container.time_gradient
+        psi = dl1_container.psi.to(u.rad).value
+        if v < 0:
+            if psi >= 0:
```

Comment:
Why are you changing psi here?

Reply:
A choice was made in the original code implementation ( before I joined the project) to use a positive time_gradient and a 2 pi rotation for psi. My interpretation is that the sign of the time_gradient is not physical.

Reply:
Makes sense! Should be explained in a comment...